File: tooling/Benchmark-Forecasting/forecasting/history_fingerprint.py
from __future__ import annotations

import hashlib
import re
from datetime import date, datetime, timezone
from typing import Any, Mapping

DATE_ONLY_PATTERN = re.compile(r"^(\d{4})-(\d{2})-(\d{2})$")
NAIVE_TIMESTAMP_PATTERN = re.compile(r"^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d{1,3}))?$")


def _format_utc_instant(value: datetime) -> str:
    return value.astimezone(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def _parse_iso_like_instant(value: str, label: str) -> datetime:
    trimmed = value.strip()
    if trimmed == "":
        raise ValueError(f"{label} must be a non-empty ISO date or timestamp.")

    date_only_match = DATE_ONLY_PATTERN.fullmatch(trimmed)
    if date_only_match is not None:
        parsed_date = date.fromisoformat(trimmed)
        return datetime(parsed_date.year, parsed_date.month, parsed_date.day, tzinfo=timezone.utc)

    naive_match = NAIVE_TIMESTAMP_PATTERN.fullmatch(trimmed)
    if naive_match is not None:
        year, month, day, hour, minute, second, milliseconds = naive_match.groups()
        return datetime(
            int(year),
            int(month),
            int(day),
            int(hour),
            int(minute),
            int(second),
            int((milliseconds or "").ljust(3, "0")[:3]) * 1000,
            tzinfo=timezone.utc,
        )

    normalized = trimmed.replace("Z", "+00:00")
    try:
        parsed = datetime.fromisoformat(normalized)
    except ValueError as exc:
        raise ValueError(f"{label} must be a valid ISO date or timestamp, received {value}.") from exc

    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)

    return parsed.astimezone(timezone.utc)


def normalize_monthly_period_identity(value: str, label: str) -> str:
    parsed = _parse_iso_like_instant(value, label)
    if (
        parsed.day != 1
        or parsed.hour != 0
        or parsed.minute != 0
        or parsed.second != 0
        or parsed.microsecond != 0
    ):
        raise ValueError(
            f"{label} must identify a canonical MONTHLY period using the first day of the month at 00:00:00.000Z, received {value}."
        )

    return _format_utc_instant(parsed)


def normalize_source_observed_at_identity(value: str, label: str) -> str:
    return _format_utc_instant(_parse_iso_like_instant(value, label))


def build_history_fingerprint(history: Mapping[str, Any]) -> str:
    normalized_points = []
    for point in history["points"]:
        source_observed_at = point.get("sourceObservedAt")
        normalized_points.append(
            {
                "date": normalize_monthly_period_identity(str(point["date"]), "Forecast history point date"),
                "value": point.get("value"),
                "sourceObservedAt": None
                if source_observed_at in (None, "")
                else normalize_source_observed_at_identity(str(source_observed_at), "Forecast history point sourceObservedAt"),
            }
        )

    normalized_points.sort(key=lambda point: point["date"])

    segments = [
        str(history["seriesId"]),
        str(history["frequency"]),
        normalize_monthly_period_identity(str(history["start"]), "Forecast history start"),
        normalize_monthly_period_identity(str(history["end"]), "Forecast history end"),
        str(history["observations"]),
    ]

    canonicalization = history.get("canonicalization")
    if isinstance(canonicalization, Mapping):
        segments.append(str(canonicalization["method"]))
        segments.append(str(canonicalization["version"]))

    for point in normalized_points:
        segment = f"{point['date']}={'null' if point['value'] is None else point['value']}"
        if point["sourceObservedAt"] is not None:
            segment = f"{segment}@{point['sourceObservedAt']}"
        segments.append(segment)

    return hashlib.sha256("\n".join(segments).encode("utf-8")).hexdigest()

File: apps/sg-runtime/lib/forecast/history-fingerprint.ts
import { createHash } from 'node:crypto'

import {
  createCalendarTargetPeriod,
  normalizeForecastSourceFrequency,
  type ForecastCadence,
  type ForecastTargetCadence,
} from '@/lib/forecast/cadence'
import { buildForecastArtifactCadenceIdentity } from '@/lib/forecast/identity'

type ForecastHistoryCanonicalization = {
  method: string
  version: string
}

type ForecastHistoryPoint = {
  date: string
  value: number | null
  sourceObservedAt?: string | null
}

export type ForecastHistoryFingerprintInput = {
  seriesId: string
  frequency: string
  cadence?: ForecastCadence
  start: string
  end: string
  observations: number
  canonicalization?: ForecastHistoryCanonicalization | null
  points: ForecastHistoryPoint[]
}

const DATE_ONLY_PATTERN = /^(\d{4})-(\d{2})-(\d{2})$/
const NAIVE_TIMESTAMP_PATTERN = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d{1,3}))?$/

function padMilliseconds(value?: string) {
  return (value ?? '').padEnd(3, '0').slice(0, 3)
}

function parseIsoLikeInstant(value: string, label: string) {
  const trimmed = value.trim()

  const dateOnlyMatch = DATE_ONLY_PATTERN.exec(trimmed)
  if (dateOnlyMatch) {
    const [, year, month, day] = dateOnlyMatch
    const parsed = new Date(Date.UTC(Number(year), Number(month) - 1, Number(day)))
    if (Number.isNaN(parsed.getTime())) {
      throw new Error(`${label} must be a valid ISO date, received ${value}.`)
    }
    return parsed
  }

  const naiveTimestampMatch = NAIVE_TIMESTAMP_PATTERN.exec(trimmed)
  if (naiveTimestampMatch) {
    const [, year, month, day, hour, minute, second, milliseconds] = naiveTimestampMatch
    const parsed = new Date(
      Date.UTC(
        Number(year),
        Number(month) - 1,
        Number(day),
        Number(hour),
        Number(minute),
        Number(second),
        Number(padMilliseconds(milliseconds)),
      ),
    )
    if (Number.isNaN(parsed.getTime())) {
      throw new Error(`${label} must be a valid ISO timestamp, received ${value}.`)
    }
    return parsed
  }

  const parsed = new Date(trimmed)
  if (Number.isNaN(parsed.getTime())) {
    throw new Error(`${label} must be a valid ISO date or timestamp, received ${value}.`)
  }

  return parsed
}

export function normalizeMonthlyPeriodIdentity(value: string, label: string) {
  const parsed = parseIsoLikeInstant(value, label)

  if (
    parsed.getUTCDate() !== 1
    || parsed.getUTCHours() !== 0
    || parsed.getUTCMinutes() !== 0
    || parsed.getUTCSeconds() !== 0
    || parsed.getUTCMilliseconds() !== 0
  ) {
    throw new Error(
      `${label} must identify a canonical MONTHLY period using the first day of the month at 00:00:00.000Z, received ${value}.`,
    )
  }

  return parsed.toISOString()
}

export function normalizeForecastPeriodIdentity(
  value: string,
  cadence: ForecastTargetCadence,
  label: string,
) {
  const parsed = parseIsoLikeInstant(value, label)
  const period = createCalendarTargetPeriod(parsed, cadence)

  if (parsed.getTime() !== period.start.getTime()) {
    throw new Error(
      `${label} must identify the canonical ${cadence} period start at 00:00:00.000Z, received ${value}.`,
    )
  }

  return parsed.toISOString()
}

export function normalizeSourceObservedAtIdentity(value: string, label: string) {
  return parseIsoLikeInstant(value, label).toISOString()
}

export function buildForecastHistoryFingerprint(history: ForecastHistoryFingerprintInput) {
  const hash = createHash('sha256')
  const targetCadence = history.cadence?.targetCadence
  if (
    history.cadence
    && normalizeForecastSourceFrequency(history.frequency) !== history.cadence.targetCadence
  ) {
    throw new Error(
      `Forecast history frequency ${history.frequency} must match target cadence ${history.cadence.targetCadence}.`,
    )
  }
  const normalizePeriod = (value: string, label: string) => targetCadence
    ? normalizeForecastPeriodIdentity(value, targetCadence, label)
    : normalizeMonthlyPeriodIdentity(value, label)
  const normalizedPoints = [...history.points]
    .map((point) => ({
      date: normalizePeriod(point.date, 'Forecast history point date'),
      value: point.value,
      sourceObservedAt: point.sourceObservedAt
        ? normalizeSourceObservedAtIdentity(point.sourceObservedAt, 'Forecast history point sourceObservedAt')
        : null,
    }))
    .sort((left, right) => left.date.localeCompare(right.date))

  hash.update(history.seriesId)
  if (history.cadence) {
    hash.update('\n')
    hash.update(buildForecastArtifactCadenceIdentity(history.cadence))
  }
  hash.update('\n')
  hash.update(history.frequency)
  hash.update('\n')
  hash.update(normalizePeriod(history.start, 'Forecast history start'))
  hash.update('\n')
  hash.update(normalizePeriod(history.end, 'Forecast history end'))
  hash.update('\n')
  hash.update(String(history.observations))

  if (history.canonicalization) {
    hash.update('\n')
    hash.update(history.canonicalization.method)
    hash.update('\n')
    hash.update(history.canonicalization.version)
  }

  for (const point of normalizedPoints) {
    hash.update('\n')
    hash.update(point.date)
    hash.update('=')
    hash.update(point.value === null ? 'null' : String(point.value))
    if (point.sourceObservedAt) {
      hash.update('@')
      hash.update(point.sourceObservedAt)
    }
  }

  return hash.digest('hex')
}